### ISSUE TITLE
Tests do not save answers for multiple select problems on page change.

### ISSUE THREAD

**Issue by somiaj** [openwebwork/webwork2#1996]
When a test has multiple pages and a problem is using a multiple select / checkbox grader, the answers are not saved when switching between pages. Here is the problem I'm testing with:

```
DOCUMENT();      
loadMacros(
    'PGstandard.pl',
    'PGML.pl',
    'PGchoicemacros.pl',
    'PGcourse.pl'
);

$mc = new_checkbox_multiple_choice();
$mc->qa('Select True Statements:', 'True1', 'True2');
$mc -> extra('False1', 'False2', 'False3');

BEGIN_PGML
[@ $mc->print_q() @]*
    [@ ANS(checkbox_cmp($mc->correct_ans())); $mc->print_a() @]*

END_PGML
ENDDOCUMENT();
```

I tested using `parserCheckboxList.pl`, and noticed the same problem too.

**Comment by somiaj** [comment]
After a little more testing, it might be saving the answers, but is not able to use the saved answers to select the associated checkboxes in tests. I noticed that if I go to the MC problem on the test, select the right answer and grade the test, I get credit and see the right checkboxes selected. But if I leave that page, then come back to it, no answers are selected anymore.

**Comment by somiaj** [comment]
Further investigation: The answers are saved correctly as a stringified array, `["B","D","E"]` for example, in the database.  But this array is then stored in a hidden field in tests when switching pages.  Storing the past answer in a hidden field assumes that data is a string, which then turns the data into a string like `ARRAY(0x559af59c99b0)`. At this point when the past answer is pulled from the hidden field it is now the string `ARRAY(0x559af59c99b0)`, which then gets stored in the database, and since it is malformed the answers are now lost.

I have tracked the issue down to line 599ish of `templates/ContentGenerator/GatewayQuiz.html.ep`.  At this point if the data is an array it needs to be put into the form in such a way it can be recovered as an array later. I've tried a few attempts but no success. I think I need to also ensure that the data gets transformed into an array correctly when it is pulled back out of the hidden field.

**Comment by somiaj** [comment]
Actually I think I have a fix, but due to @drgrice1 previous updates with checkboxes, unsure if this is valid.

```diff
diff --git a/templates/ContentGenerator/GatewayQuiz.html.ep b/templates/ContentGenerator/GatewayQuiz.html.ep
index d109497cf..51a8d6f14 100644
--- a/templates/ContentGenerator/GatewayQuiz.html.ep
+++ b/templates/ContentGenerator/GatewayQuiz.html.ep
@@ -596,7 +596,10 @@
                                % # Print out hidden fields with the current last answers.
                                % my $curr_prefix = 'Q' . sprintf('%04d', $problem_numbers->[ $probOrder->[$i] ]) . '_';
                                % for my $curr_field (grep {/^(?!previous).*$curr_prefix/} keys %{ $c->{formFields} }) {
-                                       % for (split(/\0/, $c->{formFields}{$curr_field} // '')) {
+                                       % my $data = ref($c->{formFields}{$curr_field}) eq 'ARRAY'
+                                               % ? join("\0", @{ $c->{formFields}{$curr_field} })
+                                               % : $c->{formFields}{$curr_field};
+                                       % for (split(/\0/, $data // '')) {
                                                <%= hidden_field $curr_field => $_ =%>
                                        % }
                                % }
```

@drgrice1 I can make a pull request with this fix if you think it is valid (it works in my limited testing so far).

**Comment by drgrice1** [comment]
I just removed the null byte hack.  It doesn't need to go back in again.  Pull request #1999 fixes this the correct way.

**Comment by somiaj** [comment]
Thanks for the proper fix, I wasn't sure if that null byte hack was still be used elsewhere, but had a feeling my fix wasn't proper.

**Comment by drgrice1** [comment]
PG still supports it if the front end utilizes it, but the front end shouldn't do so.  Webwork2 no longer uses the null byte hack.  I missed this instance.  The standalone renderer still needs to be updated for this I think.

**Comment by somiaj** [comment]
Fixed in #1999
